File: src/main/java/lrusso96/simplebiblio/core/SimpleBiblio.java
package lrusso96.simplebiblio.core;

import lrusso96.simplebiblio.core.providers.feedbooks.FeedbooksBuilder;
import lrusso96.simplebiblio.core.providers.libgen.LibraryGenesisBuilder;
import lrusso96.simplebiblio.core.providers.standardebooks.StandardEbooks;
import lrusso96.simplebiblio.exceptions.BiblioException;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

public class SimpleBiblio {

    final Logger logger;
    private final Set<Provider> providers;

    SimpleBiblio(Set<Provider> providers, Logger logger) {
        this.logger = logger;
        if (providers == null || providers.isEmpty())
            this.providers = setDefaultProviders();
        else
            this.providers = providers;
    }

    private Set<Provider> setDefaultProviders() {
        Set<Provider> basic = new HashSet<>();
        basic.add(new LibraryGenesisBuilder(this).build());
        basic.add(new FeedbooksBuilder(this).build());
        basic.add(new StandardEbooks(this));
        return basic;
    }

    public List<Ebook> searchAll(String query) {
        List<Ebook> ebooks = new ArrayList<>();
        //todo: support for parallel search
        for (Provider provider : providers) {
            try {
                ebooks.addAll(provider.search(query));
            } catch (BiblioException e) {
                log(e.getMessage());
            }
        }
        return ebooks;
    }

    public List<Ebook> getAllRecent() {
        List<Ebook> ebooks = new ArrayList<>();
        //todo: support for parallel search
        for (Provider provider : providers) {
            try {
                ebooks.addAll(provider.getRecent());
            } catch (BiblioException e) {
                log(e.getMessage());
            }
        }
        return ebooks;
    }

    public List<Ebook> getAllPopular() {
        List<Ebook> ebooks = new ArrayList<>();
        //todo: support for parallel search
        for (Provider provider : providers) {
            try {
                ebooks.addAll(provider.getPopular());
            } catch (BiblioException e) {
                log(e.getMessage());
            }
        }
        return ebooks;
    }

    private void log(String str) {
        if (logger != null) logger.log(Level.SEVERE, str);
    }

    public int getMaxTries() {
        return 3;
    }
}

File: src/main/java/lrusso96/simplebiblio/core/SimpleBiblio.kt
package lrusso96.simplebiblio.core

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import lrusso96.simplebiblio.core.providers.feedbooks.Feedbooks
import lrusso96.simplebiblio.core.providers.libgen.LibraryGenesis
import lrusso96.simplebiblio.core.providers.standardebooks.StandardEbooks
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

class SimpleBiblio private constructor(private val scope : CoroutineScope, private val providers: Set<Provider>) {

    suspend fun searchAll(query: String): List<Ebook> {
        val channel = Channel<List<Ebook>>()
        val ebooks: MutableList<Ebook> = ArrayList()
        providers.forEach {
            scope.launch {
                channel.send(it.search(query))
            }
            ebooks.addAll(channel.receive())
        }
        return ebooks
    }

    suspend fun getAllRecent(): List<Ebook> {
        val channel = Channel<List<Ebook>>()
        val ebooks: MutableList<Ebook> = ArrayList()
        providers.forEach {
            scope.launch {
                channel.send(it.getRecent())
            }
            ebooks.addAll(channel.receive())
        }
        return ebooks
    }


    suspend fun getAllPopular(): List<Ebook> {
        val channel = Channel<List<Ebook>>()
        val ebooks: MutableList<Ebook> = ArrayList()
        providers.forEach {
            scope.launch {
                channel.send(it.getPopular())
            }
            ebooks.addAll(channel.receive())
        }
        return ebooks
    }


    data class Builder(
            val scope : CoroutineScope = CoroutineScope(Dispatchers.IO),
            var providers: MutableSet<Provider> = HashSet()
    ) {

        private fun defaultProviders(): MutableSet<Provider> {
            logger.info { "using default providers. See doc for more information" }
            val basic: MutableSet<Provider> = HashSet()
            basic.add(LibraryGenesis.Builder().build())
            basic.add(Feedbooks.Builder().build())
            basic.add(StandardEbooks())
            return basic
        }

        fun addProvider(provider: Provider) = apply { this.providers.add(provider) }

        fun build(): SimpleBiblio {
            if (providers.isEmpty())
                providers = defaultProviders()
            return SimpleBiblio(scope, providers)
        }
    }
}